components/tikv_util/src/logger/file_log.rs

```diff
@@ -1,67 +1,264 @@
+}
+
+impl<N> RotatingFileLoggerBuilder<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    pub fn new(path: impl AsRef<Path>, rename: N) -> Self {
+        RotatingFileLoggerBuilder {
+            path: path.as_ref().to_path_buf(),
+            rotators: vec![],
+            rename,
+        }
+    }
+
+    pub fn add_rotator<R: 'static + Rotator>(mut self, rotator: R) -> Self {
+        if rotator.is_enabled() {
+            self.rotators.push(Box::new(rotator));
+        }
+        self
+    }
+
+    pub fn build(mut self) -> io::Result<RotatingFileLogger<N>> {
+        let file = open_log_file(&self.path)?;
+
+        for rotator in self.rotators.iter_mut() {
+            rotator.prepare(&file)?;
+        }
+
+        Ok(RotatingFileLogger {
+            rotators: self.rotators,
+            path: self.path,
+            rename: self.rename,
             file,
         })
     }
+}
 
-    fn should_rotate(&mut self) -> bool {
-        Utc::now() > self.next_rotation_time
+impl<N> Write for RotatingFileLogger<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    fn write(&mut self, bytes: &[u8]) -> io::Result<usize> {
+        // Updates all roators' states.
+        for rotator in self.rotators.iter_mut() {
+            rotator.update_in_write(bytes)?;
+        }
+        self.file.write(bytes)
     }
 
-    fn rename(&self) -> io::Result<PathBuf> {
+    fn flush(&mut self) -> io::Result<()> {
+        for rotator in self.rotators.iter() {
+            if rotator.should_rotate()? {
+                let new_path = (self.rename)(&self.path)?;
+                rotator.rotate(&mut self.file, &self.path, &new_path)?;
+
+                // Updates all roators' states.
+                for rotator in self.rotators.iter_mut() {
+                    rotator.update_in_flush()?;
+                }
+
+                break;
+            }
+        }
+        self.file.flush()
+    }
+}
+
+impl<N> Drop for RotatingFileLogger<N>
+where
+    N: 'static + Send + Fn(&Path) -> io::Result<PathBuf>,
+{
+    fn drop(&mut self) {
+        let _ = self.file.flush();
+    }
+}
+
+pub struct RotateByTime {
+    rotation_timespan: ReadableDuration,
+    next_rotation_time: Option<SystemTime>,
+}
+
+impl RotateByTime {
+    pub fn new(rotation_timespan: ReadableDuration) -> Self {
+        Self {
+            rotation_timespan,
+            next_rotation_time: None,
+        }
+    }
+
+    fn next_rotation_time(begin: SystemTime, duration: Duration) -> io::Result<SystemTime> {
+        begin
+            .checked_add(duration)
+            .ok_or_else(|| Error::new(ErrorKind::Other, "Next rotation time is out of range."))
+    }
+}
+
+impl Rotator for RotateByTime {
+    fn is_enabled(&self) -> bool {
+        !self.rotation_timespan.is_zero()
+    }
+
+    fn prepare(&mut self, file: &File) -> io::Result<()> {
+        self.next_rotation_time = Some(Self::next_rotation_time(
+            file.metadata()?.modified()?,
```

Comment:
next rotation time should be based on creation time instead of modified time?

Reply:
Done